jmespath-core/src/main/java/io/burt/jmespath/function/NotNullFunction.java

```diff
@@ -0,0 +1,21 @@
+package io.burt.jmespath.function;
+
+import java.util.List;
+
+import io.burt.jmespath.Adapter;
+
+public class NotNullFunction extends JmesPathFunction {
+  public NotNullFunction() {
+    super(1, Integer.MAX_VALUE);
+  }
+
+  @Override
+  protected <T> T internalCall(Adapter<T> adapter, List<ExpressionOrValue<T>> arguments) {
+    for (ExpressionOrValue<T> x : arguments) {
+      if (!adapter.isNull(x.value())) {
+        return x.value();
+      }
```

Comment:
It's indeed a problem throughout, I've added (though not yet pushed) tests for what happens when you pass an expression into every parameter to every function. This really needs to be dealt with on a generic level, because there's so much room for small differences in type checking between the functions now.